decompiler/special.py

```diff
@@ -0,0 +1,103 @@
+# Separated from unrpyc.py
+
+from . import magic
+
+
+# these named classes need some special handling for us to be able to reconstruct ren'py ASTs from pickles
+SPECIAL_CLASSES = [set, frozenset]
+
+
+# ren'py _annoyingly_ enables fix_imports even in ren'py v8,and still defaults to pickle protocol 2.
+# so set/frozenset get mapped to the wrong location (__builtins__ instead of builtins)
+# we don't want to enable that option as we want control over what the pickler is allowed to
+# unpickle
+# so here we define some proxies
+class oldset(set):
+    __module__ = "__builtin__"
+
+
+oldset.__name__ = "set"
+SPECIAL_CLASSES.append(oldset)
+
+
+class oldfrozenset(frozenset):
+    __module__ = "__builtin__"
+
+
+oldfrozenset.__name__ = "frozenset"
+SPECIAL_CLASSES.append(oldfrozenset)
+
+
+@SPECIAL_CLASSES.append
+class PyExpr(magic.FakeStrict, str):
+    __module__ = "renpy.ast"
+
+    def __new__(cls, s, filename, linenumber, py=None):
+        self = str.__new__(cls, s)
+        self.filename = filename
+        self.linenumber = linenumber
+        self.py = py
+        return self
+
+    def __getnewargs__(self):
+        if self.py is not None:
+            return str(self), self.filename, self.linenumber, self.py
+        else:
+            return str(self), self.filename, self.linenumber
+
+
+@SPECIAL_CLASSES.append
+class PyCode(magic.FakeStrict):
+    __module__ = "renpy.ast"
+
+    def __setstate__(self, state):
+        if len(state) == 4:
+            (_, self.source, self.location, self.mode) = state
+            self.py = None
+        else:
+            (_, self.source, self.location, self.mode, self.py) = state
+        self.bytecode = None
+
+
+@SPECIAL_CLASSES.append
+class Sentinel(magic.FakeStrict, object):
+    __module__ = "renpy.object"
+
+    def __new__(cls, name):
+        obj = object.__new__(cls)
+        obj.name = name
+        return obj
+
+
+# These appear in the parsed contents of user statements.
+@SPECIAL_CLASSES.append
+class RevertableList(magic.FakeStrict, list):
+    __module__ = "renpy.revertable"
+
+    def __new__(cls):
+        return list.__new__(cls)
+
+
+@SPECIAL_CLASSES.append
+class RevertableDict(magic.FakeStrict, dict):
+    __module__ = "renpy.revertable"
+
+    def __new__(cls):
+        return dict.__new__(cls)
+
+
+@SPECIAL_CLASSES.append
+class RevertableSet(magic.FakeStrict, set):
+    __module__ = "renpy.revertable"
+
+    def __new__(cls):
+        return set.__new__(cls)
+
+    def __setstate__(self, state):
+        if isinstance(state, tuple):
+            self.update(state[0].keys())
+        else:
+            self.update(state)
+
+
+class_factory = magic.FakeClassFactory(SPECIAL_CLASSES, magic.FakeStrict)
```

Comment:
Instead of just exposing class_factory as a global here, wouldn't it make more sense to just provide a function here that actually handles the unpickling, i.e.

```python
CLASS_FACTORY = magic.FakeClassFactory(SPECIAL_CLASSES, magic.FakeStrict)

def unpickle(buffer: bytes):
    return magic.safe_loads(buffer, CLASS_FACTORY, {"_ast", "collections"})
```

That way, all information on how to unpickle ren'py pickles can be contained in this file, and other files will not need to import `magic` as well. For that you'd need to also move the magic.fake_package("renpy") call into this, and ensure that this gets loaded before any ren'py imports (i.e. import the module at the start of decompiler)

Reply:
Did not see this way. I like it.